File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

n = int(input())

for i in range(n):
    s = input().rstrip()
    st = []
    check = True

    for j in s:
        if j=='(':
            st.append(j)
        else:
            if st and st[-1]=='(':
                st.pop()

            else:
                check = False
                break
            
    if not st and check:
        print('YES')
    else:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;
import java.io.*;


public class Main{
    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	
    	int n = Integer.parseInt(br.readLine());
    	Stack<Character> stack = new Stack<>();
    	StringBuilder sb = new StringBuilder();
    	
    	for (int i=0;i<n;i++) {
    		String answer = "YES";
    		String input = br.readLine();
    		stack.clear();
    		
    		for (char c : input.toCharArray()) {
    			
    			if (c=='(') {
    				stack.push(c);
    			} else if (stack.isEmpty()) {
    				answer = "NO";
    				break;
    			} else {
    				stack.pop();
    			}
    		}
    		
    		if (!stack.isEmpty()) {
    			answer = "NO";
    		}
    		sb.append(answer).append("\n");
    	}
    	
    	System.out.println(sb);
    }
}